src/sleep.rs

```diff
@@ -0,0 +1,763 @@
+        Ok(())
+    }
+}
+
+/// Will wake up the CPU based on changes in RTC GPIO pins. Is available for both light and
+/// deep sleep. However, for light sleep, the GpioWakeup struct offers more flexibility.
+/// Pullup and pulldown can be enabled for each pin. This settings will be applied when
+/// deep sleep is enabled.
+#[cfg(any(esp32, esp32s2, esp32s3))]
+pub struct RtcWakeup<P>
+where
+    P: RtcWakeupPins,
+{
+    pub pins: P,
+    pub wake_level: RtcWakeLevel,
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+impl<P> RtcWakeup<P>
+where
+    P: RtcWakeupPins,
+{
+    fn mask(&self) -> u64 {
+        let mut m: u64 = 0;
+        for pin in self.pins.iter() {
+            m |= 1 << pin;
+        }
+        m
+    }
+
+    fn apply(&self) -> Result<(), EspError> {
+        #[cfg(any(esp32, esp32s3))]
+        esp!(unsafe {
+            esp_sleep_pd_config(
+                esp_sleep_pd_domain_t_ESP_PD_DOMAIN_RTC_PERIPH,
+                esp_sleep_pd_option_t_ESP_PD_OPTION_ON,
+            )
+        })?;
+
+        esp!(unsafe { esp_sleep_enable_ext1_wakeup(self.mask(), self.wake_level.mode()) })?;
+
+        Ok(())
+    }
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+impl<P> fmt::Debug for RtcWakeup<P>
+where
+    P: RtcWakeupPins,
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "RtcWakeup {{ pins: [")?;
+
+        for pin in self.pins.iter() {
+            write!(f, "{:?} ", pin)?;
+        }
+        write!(
+            f,
+            "] mask: {:#x} wake_level: {:?} }}",
+            self.mask(),
+            self.wake_level
+        )
+    }
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum RtcWakeLevel {
+    AllLow,
+    AnyHigh,
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+impl RtcWakeLevel {
+    pub fn mode(&self) -> u32 {
+        match self {
+            RtcWakeLevel::AllLow => esp_sleep_ext1_wakeup_mode_t_ESP_EXT1_WAKEUP_ALL_LOW,
+            RtcWakeLevel::AnyHigh => esp_sleep_ext1_wakeup_mode_t_ESP_EXT1_WAKEUP_ANY_HIGH,
+        }
+    }
+}
+#[cfg(any(esp32, esp32s2, esp32s3))]
+pub trait RtcPinDriver {}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+impl<'d, P, M> RtcPinDriver for PinDriver<'d, P, M>
+where
+    P: RTCPin + InputPin + 'd,
+    M: RTCMode + 'd,
+{
+}
+
+pub trait RtcWakeupPinTrait {
+    type PinDriver;
+    fn pin(&self) -> i32;
+}
+
+pub struct DummyRtcWakeupPin;
+impl RtcWakeupPinTrait for DummyRtcWakeupPin {
+    type PinDriver = DummyRtcWakeupPin;
+    fn pin(&self) -> i32 {
+        0
+    }
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+pub trait RtcWakeupPins {
+    type WakeupPin: RtcWakeupPinTrait;
+    type Iterator<'a>: Iterator<Item = i32>
+    where
+        Self: 'a;
+
+    fn iter(&self) -> Self::Iterator<'_>;
+}
+#[cfg(not(any(esp32, esp32s2, esp32s3)))]
+pub trait RtcWakeupPins {}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+impl<P> RtcWakeupPins for P
+where
```

Comment:
Simplify to `where P: PinDriver<...>`

Reply:
I think this fails since PinDriver is a struct not a trait, or did I misunderstand what you were suggesting?